src/idl2json_cli/src/tests.rs

```diff
@@ -1,19 +1,129 @@
+//! Tests for idl2json_cli.
+#![warn(missing_docs)]
+#![allow(clippy::panic)]
+#![allow(clippy::unwrap_used)]
+#![allow(clippy::expect_used)]
+
 use super::{main, Args};
-use anyhow::{anyhow, Context};
+use anyhow::anyhow;
+use std::path::Path;
 
 #[test]
 fn simple_conversion_should_be_correct() {
-   struct TestVector {
-     stdin: &'static str, stdout: &'static str
-   }
-   let args = Args::default();
+    struct TestVector {
+        stdin: &'static str,
+        stdout: &'static str,
+    }
+    let args = Args::default();
+    let vectors = [
+        TestVector {
+            stdin: "()",
+            stdout: "",
+        },
+        TestVector {
+            stdin: "( record {} )",
+            stdout: "{}",
+        },
+        TestVector {
+            stdin: "( record {}, record {} )",
+            stdout: "{}\n{}",
+        },
+    ];
+    for vector in vectors {
+        let out = main(&args, vector.stdin)
+            .map_err(|e| anyhow!("Failed to parse: {} due to: {e}", vector.stdin))
+            .unwrap();
+        assert_eq!(vector.stdout, &out)
+    }
+}
+
+/// Returns the absolute path to a file in the samples directory.
+macro_rules! sample_file {
+    ($file:literal) => {
+        Path::new(&format!(
+            "{}/../../samples/{}",
+            env!("CARGO_MANIFEST_DIR"),
+            $file
+        ))
+        .to_path_buf()
+    };
+}
+/// Constructs idl2json_cli arguments using a type and a did file in the samples directory.
+macro_rules! typed_arg {
+    ($did_file:literal, $typ:literal) => {
+        Args {
+            did: Some(sample_file!($did_file)),
+            typ: Some($typ.to_string()),
+        }
+    };
+}
+
+#[test]
+fn conversion_with_options_should_be_correct() {
+    #[derive(Debug)]
+    struct TestVector {
+        stdin: &'static str,
+        args: Args,
+        stdout: &'static str,
+    }
+    let vectors = [
+        // On the command line we should see:
+        // $ echo "(record{canister_creation_cycles_cost= opt 999;})" | didc encode | didc decode | tee /dev/stderr | idl2json --did samples/internet_identity.did  --typ InternetIdentityInit
```

Comment:
Is `--typ` a typo?

Reply:
It's accurate.  `type` is a keyword in Rust.  It's possible to work around that but I don't believe it's worth it. So idl2json uses typ.